src/include/hash_ring.hpp

```diff
@@ -48,91 +48,101 @@ class HashRing : public ConsistentHashMap<ServerThread, H> {
     return unique_servers;
   }
 
+  bool insert_to_hash_ring(Address ip, unsigned tid) {
+    bool succeed;
+    for (unsigned virtual_num = 0; virtual_num < kVirtualThreadNum;
+         virtual_num++) {
+      succeed = HashRing::insert(ServerThread(ip, tid, virtual_num)).second;
+    }
+    return succeed;
+  }
+
+  void remove_from_hash_ring(Address ip, unsigned tid) {
+    for (unsigned virtual_num = 0; virtual_num < kVirtualThreadNum;
+         virtual_num++) {
+      HashRing::erase(ServerThread(ip, tid, virtual_num));
+    }
+  }
+
  private:
   std::unordered_set<ServerThread, ThreadHash> unique_servers;
 };
 
 typedef HashRing<GlobalHasher> GlobalHashRing;
 typedef HashRing<LocalHasher> LocalHashRing;
 
-template <typename H>
-bool insert_to_hash_ring(H& hash_ring, Address ip, unsigned tid) {
-  bool succeed;
-
-  for (unsigned virtual_num = 0; virtual_num < kVirtualThreadNum;
-       virtual_num++) {
-    succeed = hash_ring.insert(ServerThread(ip, tid, virtual_num)).second;
-  }
+class HashRingUtilInterface {
+ public:
+  virtual ServerThreadSet get_responsible_threads(
+      Address respond_address, const Key& key, bool metadata,
+      std::unordered_map<unsigned, GlobalHashRing>& global_hash_ring_map,
+      std::unordered_map<unsigned, LocalHashRing>& local_hash_ring_map,
+      std::unordered_map<Key, KeyInfo>& placement, SocketCache& pushers,
+      const std::vector<unsigned>& tier_ids, bool& succeed, unsigned& seed) = 0;
+
+  virtual ServerThreadSet responsible_global(
+      const Key& key, unsigned global_rep,
+      GlobalHashRing& global_hash_ring) = 0;
+
+  virtual std::unordered_set<unsigned> responsible_local(
+      const Key& key, unsigned local_rep, LocalHashRing& local_hash_ring) = 0;
+
+  virtual ServerThreadSet get_responsible_threads_metadata(
+      const Key& key, GlobalHashRing& global_memory_hash_ring,
+      LocalHashRing& local_memory_hash_ring) = 0;
+
+  virtual void issue_replication_factor_request(
+      const Address& respond_address, const Key& key,
+      GlobalHashRing& global_memory_hash_ring,
+      LocalHashRing& local_memory_hash_ring, SocketCache& pushers,
+      unsigned& seed) = 0;
+
+  virtual std::vector<Address> get_address_from_routing(
+      UserThread& ut, const Key& key, zmq::socket_t& sending_socket,
+      zmq::socket_t& receiving_socket, bool& succeed, Address& ip,
+      unsigned& thread_id, unsigned& rid) = 0;
+
+  virtual RoutingThread get_random_routing_thread(
+      std::vector<Address>& routing_address, unsigned& seed,
+      unsigned& kRoutingThreadCount) = 0;
+};
 
-  return succeed;
-}
+class HashRingUtil : public HashRingUtilInterface {
+ public:
+  virtual ServerThreadSet get_responsible_threads(
```

Comment:
Do we need to override all the virtual method definitions here? Aren't they automatically inherited?